source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -492,40 +486,34 @@ Network::FilterStatus Filter::onData(Buffer::Instance& data, bool end_stream) {
   ENVOY_CONN_LOG(trace, "downstream connection received {} bytes, end_stream={}",
                  read_callbacks_->connection(), data.length(), end_stream);
   getStreamInfo().addBytesReceived(data.length());
-  upstream_conn_data_->connection().write(data, end_stream);
+  if (upstream_) {
+    upstream_->encodeData(data, end_stream);
+  }
   ASSERT(0 == data.length());
```

Comment:
Should this assert be moved inside `if (upstream_`)? That said, I don't think this is called unless `upstream_` is valid.

Reply:
Yeah, it caught a few bugs on the HTTP side with readDisable not being done right.  Will comment.